cpp/src/arrow/compute/kernels/scalar_cast_internal.cc

```diff
@@ -255,7 +255,12 @@ static bool CanCastFromDictionary(Type::type type_id) {
 
 void AddCommonCasts(Type::type out_type_id, OutputType out_ty, CastFunction* func) {
   // From null to this type
-  DCHECK_OK(func->AddKernel(Type::NA, {null()}, out_ty, CastFromNull));
+  ScalarKernel kernel;
+  kernel.exec = CastFromNull;
+  kernel.signature = KernelSignature::Make({null()}, out_ty);
+  kernel.null_handling = NullHandling::COMPUTED_NO_PREALLOCATE;
+  kernel.mem_allocation = MemAllocation::NO_PREALLOCATE;
```

Comment:
Not necessary for this PR, but: the correct fix here is to modify CastFromNull to write 0s into a preallocated validity bitmap (NullHandling::COMPUTED_PREALLOCATE) and write the bare minimum to data buffers. Note we can even preallocate for binary since all the strings will be empty

Reply:
Not sure, because `MakeArrayOfNull` allocates a single (sufficiently-sized) buffer of 0s for all array buffers. Preallocating in the kernel execution machinery without any specific knowledge about the kernel would allocate distinct buffers and consume more memory (though probably not that much).

Reply:
Appended to https://issues.apache.org/jira/browse/ARROW-13121